### PR TITLE
feat: multitask transformer-agnostic model

### PR DESCRIPTION
**Pull request by dimakarp1996**
Implementation of the multitask training.

### CODE REVIEW

File: deeppavlov/core/trainers/utils.py [hunk 1]
```diff
@@ -64,4 +64,7 @@ def default(self, obj):
             return int(obj)
         elif isinstance(obj, np.floating):
             return float(obj)
-        return JSONEncoder.default(self, obj)
+        elif 'NestedExample' in str(obj.__class__):
```
Comment (IgnatovFedor) [inline deeppavlov/core/trainers/utils.py:67]: Why not isinstance (yes, in this case this class should be moved to module not required additional requirements installation)? And why this is needed?

File: deeppavlov/core/common/registry.json [hunk 2]
```diff
@@ -25,6 +25,13 @@
   "line_reader": "deeppavlov.dataset_readers.line_reader:LineReader",
   "logit_ranker": "deeppavlov.models.doc_retrieval.logit_ranker:LogitRanker",
   "mask": "deeppavlov.models.preprocessors.mask:Mask",
+  "multitask_input_splitter": "deeppavlov.models.preprocessors.multitask_preprocessor:MultiTaskInputSplitter",
+  "multitask_input_joiner": "deeppavlov.models.preprocessors.multitask_preprocessor:MultiTaskInputJoiner",
+  "multitask_reader":"deeppavlov.dataset_readers.multitask_reader:MultiTaskReader",
+  "multitask_preprocessor":"deeppavlov.models.preprocessors.multitask_preprocessor:MultiTaskPreprocessor",
+  "multitask_pipeline_preprocessor":"deeppavlov.models.preprocessors.multitask_preprocessor:MultiTaskPipelinePreprocessor",
+  "multitask_bert":"deeppavlov.models.torch_bert.torch_multitask_bert:TorchMultiTaskBert",
+  "multitask_iterator":"deeppavlov.dataset_iterators.multitask_iterator:MultiTaskIterator",
```
Comment (IgnatovFedor) [inline deeppavlov/core/common/registry.json:34]: Components should be added to requirements_registry.json, if necessary.

File: deeppavlov/core/common/registry.json [hunk 3]
```diff
@@ -25,6 +25,13 @@
   "line_reader": "deeppavlov.dataset_readers.line_reader:LineReader",
   "logit_ranker": "deeppavlov.models.doc_retrieval.logit_ranker:LogitRanker",
   "mask": "deeppavlov.models.preprocessors.mask:Mask",
+  "multitask_input_splitter": "deeppavlov.models.preprocessors.multitask_preprocessor:MultiTaskInputSplitter",
+  "multitask_input_joiner": "deeppavlov.models.preprocessors.multitask_preprocessor:MultiTaskInputJoiner",
```
Comment (IgnatovFedor) [inline deeppavlov/core/common/registry.json:29]: `multitask_input_joiner` doesn't used or mentioned anywhere

File: deeppavlov/core/common/registry.json [hunk 4]
```diff
@@ -25,6 +25,13 @@
...
+  "multitask_reader":"deeppavlov.dataset_readers.multitask_reader:MultiTaskReader",
+  "multitask_preprocessor":"deeppavlov.models.preprocessors.multitask_preprocessor:MultiTaskPreprocessor",
```
Comment (IgnatovFedor) [inline deeppavlov/core/common/registry.json:31]: doesn't exist

**Review by IgnatovFedor — CHANGES_REQUESTED** [review]
Get rid of asserts: remove unnecessary, necessary ones replace with `if ...: raise ...Exception(...)`.
Get rid of unnecessary prints

File: deeppavlov/utils/glue_eval/evaluate_glue_superglue.py [hunk 5]
```diff
@@ -397,20 +368,11 @@ def get_glue_metric(task,split='test',log_dict=True,submit_dir=''):
             metric = pearsonr(predictions, labels)[0]
         else:
             metric = accuracy_score(predictions, labels)
-        print(metric)
```
Comment (dimakarp1996) [inline deeppavlov/utils/glue_eval/evaluate_glue_superglue.py]: На валидации метрика должна печататься.

File: deeppavlov/utils/glue_eval/evaluate_glue_superglue.py [hunk 6]
```diff
@@ -397,20 +368,11 @@ def get_glue_metric(task,split='test',log_dict=True,submit_dir=''):
...
-        if metric<0.5:
-            print(pred_to_true)
```
Comment (dimakarp1996) [inline deeppavlov/utils/glue_eval/evaluate_glue_superglue.py]: Это для отладки. Классы можно перепутать, это на случай, если классы перепутаны

File: deeppavlov/core/trainers/utils.py (hunk 1 above)
Comment (dimakarp1996) [reply deeppavlov/core/trainers/utils.py:67]: 1)Чтобы корректно работать с RecordNestedExample, в смысле корректно печатать их. А то был баг
2)Ты уверен, что RecordNestedExample - единственный класс в либе, содержащий NestedExample?

File: deeppavlov/dataset_readers/basic_classification_reader.py [hunk 7]
```diff
@@ -72,16 +81,23 @@ def read(self, data_path: str, url: str = None,
...
             if file_name is None:
                 continue
-            
+
             file = Path(data_path).joinpath(file_name)
             if file.exists():
                 if format == 'csv':
                     keys = ('sep', 'header', 'names')
                     options = {k: kwargs[k] for k in keys if k in kwargs}
+                    print(file)
```
[the remark this answers is not in the document]
Comment (dimakarp1996) [reply deeppavlov/dataset_readers/basic_classification_reader.py:100]: It is important to know what file is processed, for debugging purpose. For me as a developer, it was important.

File: deeppavlov/core/common/metrics_registry.json [hunk 8]
```diff
@@ -9,6 +9,7 @@
   "f1": "deeppavlov.metrics.fmeasure:round_f1",
   "f1_macro": "deeppavlov.metrics.fmeasure:round_f1_macro",
   "f1_weighted": "deeppavlov.metrics.fmeasure:round_f1_weighted",
+  "f1_weighted_with_confusion_matrix": "deeppavlov.metrics.fmeasure:round_f1_weighted_with_confusion_matrix",
```
[the remark this answers is not in the document]
Comment (dimakarp1996) [reply deeppavlov/core/common/metrics_registry.json:12]: OK, I remove it as this is the matter of another PR, https://github.com/deeppavlov/DeepPavlov/pull/1623

File: deeppavlov/core/data/simple_vocab.py [hunk 9]
```diff
@@ -89,7 +89,11 @@ def _add_tokens_with_freqs(self, tokens, freqs):
 
     def __call__(self, batch, is_top=True, **kwargs):
         if isinstance(batch, Iterable) and not isinstance(batch, str):
-            looked_up_batch = [self(sample, is_top=False) for sample in batch]
+            if all([k is None for k in batch]):
+                # for multitask dummy input. With multitask 1 batch == 1  task but fields for all tasks exists
```
[the remark this answers is not in the document]
Comment (dimakarp1996) [reply deeppavlov/core/data/simple_vocab.py]: Removed comment

File: deeppavlov/dataset_iterators/huggingface_dataset_iterator.py [hunk 10]
```diff
@@ -43,11 +43,20 @@ def preprocess(self,
         """
 
         dataset = []
-        for example in data:
+        for i in range(len(data)):  # for example in data
+            try:
+                example = data[i]
+            except Exception as e:
```
[the remark this answers is not in the document]
Comment (dimakarp1996) [reply deeppavlov/dataset_iterators/huggingface_dataset_iterator.py]: OK - that was overkill

File: deeppavlov/dataset_iterators/huggingface_dataset_iterator.py [hunk 11]
```diff
@@ -43,11 +43,20 @@ def preprocess(self,
...
+                print(f'Exception with features {features} and dataset index {i}')
```
[the remark this answers is not in the document]
Comment (dimakarp1996) [reply deeppavlov/dataset_iterators/huggingface_dataset_iterator.py]: OK - that was overkill

File: deeppavlov/core/common/registry.json (hunk 3 above)
Comment (dimakarp1996) [reply deeppavlov/core/common/registry.json]: This is an old artifact. removed

File: deeppavlov/core/common/registry.json (hunk 4 above)
Comment (dimakarp1996) [reply deeppavlov/core/common/registry.json]: Removed. It was an old artifact

File: deeppavlov/dataset_iterators/huggingface_dataset_iterator.py [hunk 12]
```diff
@@ -43,11 +43,20 @@ def preprocess(self,
...
+                raise e
             if isinstance(features, str):
                 feat = example[features]
             elif isinstance(features, list):
-                feat = tuple(example[f] for f in features)
+                if not all([f in example for f in features]):
```
[the remark this answers is not in the document]
Comment (dimakarp1996) [reply deeppavlov/dataset_iterators/huggingface_dataset_iterator.py]: ok

**Pull request merged**